# Lab 1 - Stefan Regalia

## 1. Covid-19 Dataset

## How does sex relate to outcome?

### Pseudocode:
- Import libraries and upload the dataset
- Group the dataset by sex
- Count the total values for each possible outcome for each sex
- Convert these values to a dataframe

In [86]:
# Import Libraries
import pandas as pd
import os
os.getcwd()

'/workspaces/DS-3021/02_function_basics'

In [87]:
covid = pd.read_csv('../data/COVID-19_final.csv') # Enters the data directory and reads the Covid-19 csv final
covid.head() # Prints the first five rows of the COVID-19 dataset


age     sex    country     outcome travel_history_binary  Hypertension  \
0  78.0    male      Italy    Deceased                   NaN             0   
1  61.0  female  Singapore  Discharged                   NaN             0   
2  28.0    male  Singapore  Discharged                   NaN             0   
3  56.0  female  Singapore  Discharged                   NaN             0   
4  79.0  female  Singapore  Discharged                   NaN             0   

   COPD  Diabetes  Heart Disease  Chronic Kidney Disease  Asthma  Cancer  \
0     0         0              0                       0       0       0   
1     0         0              0                       0       0       0   
2     0         0              0                       0       0       0   
3     0         0              0                       0       0       0   
4     0         0              0                       0       0       0   

   symp_fatigue  symp_acuterespsyndrome  symp_acuterespfail  symp_cough  \
0             0                       0                   0           0   
1             0                       0                   0           0   
2             0                       0                   0           0   
3             0                       0                   0           0   
4             0                       0                   0           0   

   symp_fever  symp_sorethroat  symp_septic  symp_pneu  
0           0                0            0          0  
1           0                0            0          0  
2           0                0            0          0  
3           0                0            0          0  
4           0                0            0          0

In [111]:
by_sex = covid.groupby('sex')['outcome'] # Groups the dataset by sex to determine the difference in outcome between males and females

In [117]:
sex_outcomes = pd.DataFrame(by_sex.value_counts()) # Counts the number of cases that were discharged or deceased for males and females, then turns it into a dataframe
sex_outcomes # Displays the dataframe

count
sex    outcome          
female Discharged    141
       Deceased       71
male   Discharged    189
       Deceased      132

In [118]:
def outcomes_by_sex(x):
    by_sex = covid.groupby('sex')['outcome']
    sex_outcomes = pd.DataFrame(by_sex.value_counts())

    return sex_outcomes

For females, 141/(71+141) = 66.5% of cases in this dataset survived and were discharged, while for males, 189/(189+132) = 58.9% survived and were discharged. While it may appear that females were more likely to survive COVID-19, we have not accounted for any possible confounding variables, so we can not conclude this.

## What is the relationship between age and outcome?

### Pseudocode: 
- We will break the dataset into four age groups: Child (1-18), Adult (19-40), Middle Age (41-59), and Senior (60+)
- We will then count the total different outcomes for each of the four age groups and dividing it by the length of that column to get fair proportions.
- Draw conclusions

In [ ]:
child = covid.query("age < 19")  # Filtering for people under 19
adult = covid.query("19 <= age <= 40")  # Filtering for people between 19 and 40
middle_age = covid.query("41 <= age <= 59")  # Filtering for people between 41 and 59
senior = covid.query("age >= 60")  # Filtering for of people 60 or older

In [129]:
child_outcomes = child.groupby('outcome').size()/len(child) # Calculates the proportion of each outcome for children
adult_outcomes = adult.groupby('outcome').size()/len(adult) # Calculates the proportion of each outcome for adults
middle_age_outcomes = middle_age.groupby('outcome').size()/len(middle_age) # Calculates the proportion of each outcome for middle aged people
senior_outcomes = senior.groupby('outcome').size()/len(senior) # Calculates the proportion of each outcome for seniors

In [134]:
# Prints the proportion of discharged and deceased people per age group

print(child_outcomes)
print(adult_outcomes)
print(middle_age_outcomes)
print(senior_outcomes)

outcome
Discharged    1.0
dtype: float64
outcome
Deceased      0.041667
Discharged    0.958333
dtype: float64
outcome
Deceased      0.246575
Discharged    0.753425
dtype: float64
outcome
Deceased      0.757143
Discharged    0.242857
dtype: float64


In [135]:
def age_on_outcome(x):
    child = covid.query("age < 19")  
    adult = covid.query("19 <= age <= 40")  
    middle_age = covid.query("41 <= age <= 59")  
    senior = covid.query("age >= 60")
    child_outcomes = child.groupby('outcome').size()/len(child) 
    adult_outcomes = adult.groupby('outcome').size()/len(adult) 
    middle_age_outcomes = middle_age.groupby('outcome').size()/len(middle_age) 
    senior_outcomes = senior.groupby('outcome').size()/len(senior)

    return child_outcomes, adult_outcomes, middle_age_outcomes, senior_outcomes
    

From this code, it is quite clear that there is a correlation between younger people surviving COVID-19 more than the elderly. 

## 2. Hotels Dataset

## How does the arrival month relate to the type of hotel?

### Pseudocode:
- Group by the arrival month
- Count how many of each type of hotel there are for each month
- Convert to a dataframe

In [152]:
hotels = pd.read_csv('../data/hotels.csv') # Fetch the hotels csv from the data directory
hotels.head() # Print the first five lines of the hotels directory

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

In [ ]:
by_month = hotels.groupby('arrival_date_month')['hotel'] # Grouping by arrival month at the hotel 

In [156]:
hotel_type = pd.DataFrame(by_month.value_counts()) # Counts the number of bookings per hotel type per month and converts it to a dataframe

count
arrival_date_month hotel              
April              City Hotel     7480
                   Resort Hotel   3609
August             City Hotel     8983
                   Resort Hotel   4894
December           City Hotel     4132
                   Resort Hotel   2648
February           City Hotel     4965
                   Resort Hotel   3103
January            City Hotel     3736
                   Resort Hotel   2193
July               City Hotel     8088
                   Resort Hotel   4573
June               City Hotel     7894
                   Resort Hotel   3045
March              City Hotel     6458
                   Resort Hotel   3336
May                City Hotel     8232
                   Resort Hotel   3559
November           City Hotel     4357
                   Resort Hotel   2437
October            City Hotel     7605
                   Resort Hotel   3555
September          City Hotel     7400
                   Resort Hotel   3108

In [157]:
def hotel_type_per_month(x):
    by_month = hotels.groupby('arrival_date_month')['hotel']
    hotel_type = pd.DataFrame(by_month.value_counts())

    return hotel_type

It appears that more people book city hotel year around, so there is no correlation between the month and the type of hotel booked. However, there is correlation between the month and number of bookings for both types of hotels as hotel bookings increase May-September for the summer.

## What is the relationship between lead_time and the number of adults in a booking during summer months (May-September)?

### Pseudocode:
- Filter the dataset for booking in months May-September
- Determine the number of adults in these bookings
- Calculate the mean lead_time for these bookings

In [ ]:
# Filter the dataset for booking in months May-September

summer_months = hotels[hotels['arrival_date_month'].isin(['May', 'June', 'July', 'August', 'September'])]

In [170]:
summer_adults = summer_months.groupby('adults')['lead_time'].mean() # Find the mean lead_time for all different numbers of adults on bookings.

In [174]:
lead_times = pd.DataFrame(summer_adults) # Creates a dataframe of all of the mean values
lead_times

lead_time
adults            
0       117.228916
1        90.554434
2       136.347197
3       107.651945
4        82.531915
26      343.500000
27      350.500000
40      304.000000
50      336.000000

In [173]:
def summer_lead_time_adults(x):
    summer_months = hotels[hotels['arrival_date_month'].isin(['May', 'June', 'July', 'August', 'September'])]
    summer_adults = summer_months.groupby('adults')['lead_time'].mean()

    return lead_times

Based on this new dataframe, there was not much correlation between the number of adults and lead_time in the summer months when there were only a few adults in the booking. However, bookings with many adults booked almost a year in advance.